src/libcharon/plugins/kernel_super/kernel_super_ipsec.c

```diff
@@ -3050,34 +3187,61 @@ static bool tunnel_counter_key_compare(const void *key, const void *other_key) {
 
-	const ike_super_peer_info_t* tunnel_1_peer_info = tunnel_1->super_peer_info;
-	const ike_super_peer_info_t* tunnel_2_peer_info = tunnel_2->super_peer_info;
 
 	return tunnel_1->terminator_id == tunnel_2->terminator_id &&
-			streq(tunnel_1_peer_info->get_tenant_id(tunnel_1_peer_info), tunnel_2_peer_info->get_tenant_id(tunnel_2_peer_info)) &&
-			streq(tunnel_1_peer_info->get_account_id(tunnel_1_peer_info), tunnel_2_peer_info->get_account_id(tunnel_2_peer_info)) &&
-			streq(tunnel_1_peer_info->get_vrf_name(tunnel_1_peer_info), tunnel_2_peer_info->get_vrf_name(tunnel_2_peer_info));
+			streq(tunnel_1->tenant_id, tunnel_2->tenant_id) &&
+			streq(tunnel_1->account_id, tunnel_2->account_id) &&
+			streq(tunnel_1->vrf_name, tunnel_2->vrf_name);
 }
 
 static u_int tunnel_counter_key_hash(const void *key) {
 	const struct tunnel_counter_key *tunnel_key = (const struct tunnel_counter_key*)key;
-	const ike_super_peer_info_t* const super_peer_info = tunnel_key->super_peer_info;
 
 	return chunk_hash_inc(chunk_from_thing(tunnel_key->terminator_id),
-						  chunk_hash_inc(chunk_from_str(super_peer_info->get_tenant_id(super_peer_info)),
-						  chunk_hash_inc(chunk_from_str(super_peer_info->get_account_id(super_peer_info)),
-						  chunk_hash(chunk_from_str(super_peer_info->get_vrf_name(super_peer_info))))));
+						  chunk_hash_inc(chunk_from_str(tunnel_key->tenant_id),
+						  chunk_hash_inc(chunk_from_str(tunnel_key->account_id),
+						  chunk_hash(chunk_from_str(tunnel_key->vrf_name)))));
 }
 
 
 static void destroy_tunnel_counter_key(void *value, const void *key) {
 	(void)key;
-	struct tunnel_counter_key *p_val = (struct tunnel_counter_key*)value;
-	DESTROY_IF(p_val->super_peer_info);
+
+	struct tunnel_counter_key *tunnel_key = (struct tunnel_counter_key *)value;
+	free(tunnel_key->tenant_id);
+	free(tunnel_key->account_id);
+	free(tunnel_key->vrf_name);
 	free(value);
 }
 
-static void write_terminator_tunnels_oper(struct pm_terminator_conn* const terminator_conn) {
+CALLBACK(write_terminator_tenant_account_vrf_tunnels, void,
+		 hashtable_t *ctx, const struct system_enterprise_map_iterator* iter) {
+	const char* const tenant_id = iter->lookup_params->tenant_id;
+	const char* const account_id = iter->lookup_params->account_id;
+	const char* const vrf_name = iter->lookup_params->vrf_name;
+	uint32_t term_id = iter->lookup_params->terminator_id;
+
+	struct tunnel_counter_key key = {.terminator_id = term_id,
+									 .tenant_id = tenant_id,
+									 .account_id = account_id,
+									 .vrf_name = vrf_name,
+									 .count = 0};
+	struct tunnel_counter_key *curr_key = ctx->get(ctx, &key);
+	if(NULL != curr_key) {
+		key.count = curr_key->count;
+	}
+
+	if (! oper_db_writer_set_terminator_vrf_tunnel_counter(term_id,
+														   tenant_id, account_id, vrf_name,
+														   key.count)) {
+		LOG(DBG_KNL, SL_LOG_ERR, "failed writing vrf: %s in term %u tunnel counter",
+			vrf_name, curr_key->terminator_id);
```

Comment:
`curr_key->terminator_id` will cause a NULL dereference if `curr_key` is NULL. Use `term_id` instead